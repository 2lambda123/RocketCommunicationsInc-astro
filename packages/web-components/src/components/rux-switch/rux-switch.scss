@use '../../common/functional-components/FormFieldMessage/form-field-message.scss';
:host {
    display: block;

    /**
      * @prop --switch-background-color: the Switch off color
      */

    --switch-background-color: var(--color-background-base-default);

    /**
      * @prop --switch-hover-on-color: the Switch hover on color
      */
    --switch-hover-on-color: var(--color-background-interactive-default);

    /**
      * @prop --switch-hover-off-color: the Switch hover off color
      */

    --switch-hover-off-color: var(--color-background-interactive-default);

    /** @prop --switch-on-color: the Switch on color */
    --switch-on-color: var(--color-background-interactive-default);

    /** @prop --switch-off-border-color: the Switch off border color */
    --switch-off-border-color: var(--color-border-interactive-muted);

    padding-left: 1%;
}
.rux-switch {
    position: relative;
    display: flex;
    height: 22px;
    white-space: nowrap;

    // Default Styling
    &__button {
        display: flex;
        align-items: center;
        cursor: pointer;

        // Track
        &::before {
            position: relative;
            display: flex;
            content: '';
            border-radius: var(--switch-radius-track);
            border: 1px solid;
            border-color: var(--switch-off-border-color);
            background-color: var(--switch-background-color);
            height: 10px;
            width: 30px;
            z-index: 2;
            transition: 0.1s background-color linear;
        }

        // Button
        &::after {
            position: absolute;
            content: '';
            top: 1px;
            left: -4px;
            z-index: 3;
            height: 18px;
            width: 18px;
            border-radius: var(--radius-circle);
            border: 1px solid var(--switch-hover-off-color);
            background-color: var(--switch-background-color);
            transition: 0.1s left linear, 0.1s border-color linear;
        }
    }

    &__input {
        appearance: none;
        -webkit-appearance: none;
        margin: 0;

        // Active Styling
        &:checked {
            + .rux-switch__button {
                &::before {
                    border-color: var(--switch-on-color);
                    background-color: var(--switch-on-color);
                }
                &::after {
                    left: 16px;
                    background-color: var(--switch-background-color);
                    border-color: var(--switch-hover-off-color);
                }
            }
        }

        // Disabled
        &:disabled {
            + .rux-switch__button {
                cursor: not-allowed;
                opacity: var(--disabled-opacity);
            }
        }
    }

    // :before -> Track, :after -> Button
    &:hover {
        .rux-switch__input {
            &:not(:disabled),
            &:checked:not(:disabled) {
                + .rux-switch__button {
                    &::before {
                        border-color: var(--switch-hover-off-color);
                    }
                    &::after {
                        border-color: var(--color-background-interactive-hover);
                    }
                }
            }

            &:not(:disabled) {
                + .rux-switch__button {
                    &::before {
                        background-color: var(--switch-background-color);
                    }
                }
            }

            &:checked:not(:disabled) {
                + .rux-switch__button {
                    &::before {
                        background-color: var(--switch-hover-on-color);
                    }
                }
            }
        }
    }
}

.rux-switch__label {
    margin-left: 0.75rem;
}

.rux-form-field {
    display: flex;
    flex-direction: column;
    color: var(--color-text-primary);
}

.hidden,
:host([hidden]) {
    display: none;
}
