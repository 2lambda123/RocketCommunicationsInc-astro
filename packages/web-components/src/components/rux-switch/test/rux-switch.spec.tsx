import { newSpecPage } from '@stencil/core/testing'
import { RuxSwitch } from '../rux-switch'

describe('rux-switch', () => {
    it('builds', async () => {
        const switchBtn = new RuxSwitch()
        const { switchId, checked, disabled, name, value } = switchBtn

        expect(switchBtn).toBeTruthy()
        expect({
            switchId,
            checked,
            disabled,
            name,
            value,
        }).toEqual({
            switchId: 'rux-switch-1',
            checked: false,
            disabled: false,
            errorText: undefined,
            name: '',
            value: '',
        })
    })

    it('auto-increments', () => {
        const switchBtn = new RuxSwitch()
        const { switchId } = switchBtn
        expect({ switchId }).toEqual({ switchId: 'rux-switch-2' })
    })

    it('renders', async () => {
        const page = await newSpecPage({
            components: [RuxSwitch],
            html: `<rux-switch></rux-switch>`,
        })
        expect(page.root).toMatchSnapshot()
    })

    it('renders label prop', async () => {
        const page = await newSpecPage({
            components: [RuxSwitch],
            html: `<rux-switch label="hello"></rux-switch>`,
        })
        expect(page.root).toMatchSnapshot()
    })
    it('renders label slot', async () => {
        const page = await newSpecPage({
            components: [RuxSwitch],
            html: `<rux-switch><div slot="label">hello</div></rux-switch>`,
        })
        expect(page.root).toMatchSnapshot()
    })
})
