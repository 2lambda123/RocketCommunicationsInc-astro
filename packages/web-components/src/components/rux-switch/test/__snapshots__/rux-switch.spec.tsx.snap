// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rux-switch renders 1`] = `
<rux-switch aria-checked="false" class="rux-form-field" role="switch" value="">
  <mock:shadow-root>
    <div class="rux-switch">
      <input aria-checked="false" class="rux-switch__input" id="rux-switch-3" role="switch" type="checkbox" value="">
      <label class="rux-switch__button" htmlfor="rux-switch-3">
        <span class="hidden rux-switch__label">
          <slot name="label"></slot>
        </span>
      </label>
    </div>
  </mock:shadow-root>
</rux-switch>
`;

exports[`rux-switch renders label prop 1`] = `
<rux-switch aria-checked="false" class="rux-form-field" label="hello" role="switch" value="">
  <mock:shadow-root>
    <div class="rux-switch">
      <input aria-checked="false" class="rux-switch__input" id="rux-switch-4" role="switch" type="checkbox" value="">
      <label class="rux-switch__button" htmlfor="rux-switch-4">
        <span class="rux-switch__label">
          <slot name="label">
            hello
          </slot>
        </span>
      </label>
    </div>
  </mock:shadow-root>
</rux-switch>
`;

exports[`rux-switch renders label slot 1`] = `
<rux-switch aria-checked="false" class="rux-form-field" role="switch" value="">
  <mock:shadow-root>
    <div class="rux-switch">
      <input aria-checked="false" class="rux-switch__input" id="rux-switch-5" role="switch" type="checkbox" value="">
      <label class="rux-switch__button" htmlfor="rux-switch-5">
        <span class="rux-switch__label">
          <slot name="label"></slot>
        </span>
      </label>
    </div>
  </mock:shadow-root>
  <div slot="label">
    hello
  </div>
</rux-switch>
`;
